fix/#80-yoast-import-error

nickmoreton opened the pull request:
An eror was occuring where the xml_item_key was returned as a list. This happens if there is only one wp:meta_key and wp:meta_value

nickmoreton committed:
change get_yoast_description_value method

An eror was occuring where the xml_item_key was returned as a list. This happens if there is only one wp:meta_key and wp:meta_value

nimasmi commented on wagtail_wordpress_import/importers/wordpress.py:
```diff
@@ -335,13 +335,19 @@ def get_yoast_description_value(self):
+
+        if xml_item_key in node_keys:
+            for item in self.node.get(xml_item_key):
+                if type(item) == dict:
```
**Suggestion**: use isinstance, as it covers inherited subtypes too

```suggestion
                if isinstance(item, dict):
```

nickmoreton replied:
**Question** I'm not sure how to write a test that would fail without this fix?

but I have added a new test to check a single wp:postmeta runs successfully.

nimasmi replied:
Sorry, Nick, I only just saw this comment. Are you talking about the `isinstance` fix or the single vs. multiple `<wp:postmeta>` tags fix? The former doesn't need a unit test, the latter I can look at today.

nickmoreton replied:
I was looking at it last night, the issue was when a None type is returned. I think I have a better idea for that function, shall I do that work before you review again? Also the next ticket codebase #73 for dealing with the import hooks might mean altering the process around yoast and other wp:postmeta elements.

nickmoreton replied:
Second thoughts if poss and this fix is acceptable it might be better to stay with this for now.

nimasmi reviewed (changes requested):
I'm not completely sure what error was being reported here. Since this is a bugfix, is it possible to add a testcase that fails without the fix, and is fixed by these changes?

nickmoreton committed:
add new fixture to represent a single wp:postmeta key

nickmoreton committed:
add new test for encountering a single wp:postmeta key

nickmoreton committed:
change to use isinstance

nimasmi commented on wagtail_wordpress_import/importers/wordpress.py:
```diff
@@ -335,13 +335,19 @@ def get_yoast_description_value(self):
+                        meta_key_values[0]
+                        == yoast_plugin_config()["description_key_value"]
+                    ):
+                        meta_value = meta_key_values[1] or ""
```
**Suggestion**: This looks like it does the same as:

```suggestion
        config = yoast_plugin_config()
        node_keys = self.node.keys()
        xml_item_key = config["xml_item_key"]

        for item in self.node.get(xml_item_key):
            if item[config["description_key"]] == config["description_key_value"]:
                meta_value = item[config["description_value"]] or ""
```

nickmoreton committed:
Revert "change get_yoast_description_value method"

This reverts commit 708da4edbab5ac40f4895090308bd0bc6939e7d8.

# Conflicts:
#	wagtail_wordpress_import/importers/wordpress.py

nickmoreton committed:
Revert "add new test for encountering a single wp:postmeta key"

This reverts commit ec963ad8716ac75364c464222ca1e85d21641088.

nickmoreton committed:
change expected logger values

the xml fixture has extra items to import, alter these expected values to reflect the change

nickmoreton committed:
add new test fixture for testing the Exception

save

nickmoreton committed:
add isinstance check

nickmoreton committed:
change get_yoast_description_value() to ensure meta_items is a list of dicts

nickmoreton committed:
add new tests for  get_yoast_description_value()

nimasmi commented on wagtail_wordpress_import/test/fixtures/raw_xml.xml:
```diff
@@ -252,6 +252,35 @@ Nihil hic munitissimus habendi senatus locus, nihil horum?.&lt;/blockquote&gt;</
+                <wp:meta_key>_thumbnail_id</wp:meta_key>
+                <wp:meta_value><![CDATA[43124]]></wp:meta_value>
+            </wp:postmeta>
+        </item>
```
**Question**: What role does this new fixture have? All the unit tests in https://github.com/torchbox/wagtail-wordpress-import/pull/72/files#diff-87e9fedd3c41cadd0bc524052e05b7d963c00dbd029ac32a3ef85757b3e24974R214 seem to use the `post_meta.xml` file.

nickmoreton replied:
Thanks Nick. That fixture is used in [test_wordpress](https://github.com/torchbox/wagtail-wordpress-import/blob/fix/%2380-yoast-import-error/wagtail_wordpress_import/test/tests/test_wordpress_importer.py) The fixture didn't have the single wp_post_meta tag to test.

I think I added it there before deciding to do the tests in test_wordpress_item.py. I'll have a look but rather than remove the fixture i'll add a new test for it in test_wordpress.py

nickmoreton replied:
Actually doesn't seem like test_wordpress.py is a good place for a test like that as we don't do anything with _thumbnail_id. We will in the next piece of work but not here. I'll remove it.

nickmoreton replied:
I reverted this extra fixture and the commit where the total need to be updated.

nimasmi commented on wagtail_wordpress_import/test/tests/test_wordpress_item.py:
```diff
@@ -195,3 +197,76 @@ def test_page_has_no_categories(self):
+@override_settings(
+    WAGTAIL_WORDPRESS_IMPORT_YOAST_PLUGIN_ENABLED=True,
+)
+class WordpressImporterTestsYoastEnabledSingleKey(TestCase):
```
**Suggestion**: Rename this to `WordpressImporterTestsYoastMetaDescriptions`, as it is not only testing single keys.

nickmoreton replied:
Thank Nick I change the naming here.

nimasmi reviewed (commented):
Thanks Nick, this looks like a big improvement. I have two questions, but I'm also happy with the changes so far.

nickmoreton committed:
Revert "add new fixture to represent a single wp:postmeta key"

This reverts commit 0bc4be5f7bebc4016d25409d0afc62d3c94f933c.

nickmoreton committed:
Revert "change expected logger values"

This reverts commit 648a22ee77972713bbcb3771ce8bf840c0d111d7.

nickmoreton committed:
change test class name

nimasmi reviewed (approved):
Thanks @nickmoreton
